packages/langium/src/references/scope-provider.ts

```diff
@@ -165,7 +98,12 @@ export class DefaultScopeProvider implements ScopeProvider {
      * Create a global scope filtered for the given reference type.
      */
     protected getGlobalScope(referenceType: string, _context: ReferenceInfo): Scope {
-        return new StreamScope(this.indexManager.allElements(referenceType));
+        let scope = this.globalScopeCache.get(referenceType);
+        if (!scope) {
+            scope = new MapScope(this.indexManager.allElements(referenceType));
+            this.globalScopeCache.set(referenceType, scope);
+        }
+        return scope;
```

Comment:
We should extract this cache access, either to a utility class
```typescript
return this.globalScopeCache.get(referenceType, () => new MapScope(this.indexManager.allElements(referenceType)));
```
or to another protected method
```typescript
return this.fromGlobalScopeCache(referenceType, () => new MapScope(this.indexManager.allElements(referenceType)));
```

Reply:
This will become easier once we switch to the CachingService (#1123).